In [1]:
import pandas as pd  # pandas
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

# Adatbázis séma (Diagram)

<img src="table_diagram.png">

# Adattáblák

## events

In [2]:
events = pd.read_csv("../datasets/events.csv", sep=",")

In [3]:
events.head()

event_id            device_id            timestamp  longitude  latitude
0         1    29182687948017175  2016-05-01 00:55:25     121.38     31.24
1         2 -6401643145415154744  2016-05-01 00:54:12     103.65     30.97
2         3 -4833982096941402721  2016-05-01 00:08:05     106.60     29.70
3         4 -6815121365017318426  2016-05-01 00:06:40     104.27     23.28
4         5 -5373797595892518570  2016-05-01 00:07:18     115.88     28.66

In [4]:
events.count()

event_id     3252950
device_id    3252950
timestamp    3252950
longitude    3252950
latitude     3252950
dtype: int64

## gender_age_train

In [5]:
gender_age_train = pd.read_csv("../datasets/gender_age_train.csv", sep=",")

In [6]:
gender_age_train.head()

device_id gender  age   group
0 -8076087639492063270      M   35  M32-38
1 -2897161552818060146      M   35  M32-38
2 -8260683887967679142      M   35  M32-38
3 -4938849341048082022      M   30  M29-31
4   245133531816851882      M   30  M29-31

In [7]:
gender_age_train.count()

device_id    74645
gender       74645
age          74645
group        74645
dtype: int64

In [8]:
gender_age_train.groupby('group')['device_id'].count()

group
F23-      5050
F24-26    4190
F27-28    3118
F29-32    4628
F33-42    5561
F43+      4194
M22-      7488
M23-26    9605
M27-28    5445
M29-31    7309
M32-38    9476
M39+      8581
Name: device_id, dtype: int64

In [9]:
gender_age_train.groupby('gender')['device_id'].count()

gender
F    26741
M    47904
Name: device_id, dtype: int64

## phone_brand_device_model

In [10]:
phone_brand_device_model = pd.read_csv("../datasets/phone_brand_device_model.csv", sep=",")

In [11]:
phone_brand_device_model.head()

device_id phone_brand   device_model
0 -8890648629457979026          小米             红米
1  1277779817574759137          小米           MI 2
2  5137427614288105724          三星      Galaxy S4
3  3669464369358936369       SUGAR           时尚手机
4 -5019277647504317457          三星  Galaxy Note 2

In [12]:
# a kínai márkákat lefordítjuk...

In [13]:
translate = pd.read_csv("../datasets/translate.csv", sep=",")

In [14]:
translate.head()

phone_brand phone_brand_eng
0          三星         samsung
1          天语          Ktouch
2          海信         hisense
3          联想          lenovo
4          欧比             obi

In [15]:
phone_brand_device_model= phone_brand_device_model.merge(translate,on='phone_brand',how='right')

In [16]:
phone_brand_device_model.head()

device_id phone_brand device_model phone_brand_eng
0 -8890648629457979026          小米           红米          Xiaomi
1  1277779817574759137          小米         MI 2          Xiaomi
2 -3883532755183027260          小米        MI 2S          Xiaomi
3 -5827952925479472594          小米  MI One Plus          Xiaomi
4 -5776341595519925628          小米        MI 2A          Xiaomi

## app_events

In [17]:
app_events = pd.read_csv("../datasets/app_events.csv", sep=",")

In [18]:
app_events.head()

event_id               app_id  is_installed  is_active
0         2  5927333115845830913             1          1
1         2 -5720078949152207372             1          0
2         2 -1633887856876571208             1          0
3         2  -653184325010919369             1          1
4         2  8693964245073640147             1          1

In [19]:
app_events.count()

event_id        32473067
app_id          32473067
is_installed    32473067
is_active       32473067
dtype: int64

In [20]:
app_events.groupby('event_id')['app_id'].agg('count').count()

1488096

In [21]:
# osszesen 1488096 db esemény van, ahhoz van 32473067 alkalmazás esemény 

## app_labels

In [22]:
app_labels = pd.read_csv("../datasets/app_labels.csv", sep=",")

In [23]:
app_labels.head()

app_id  label_id
0  7324884708820027918       251
1 -4494216993218550286       251
2  6058196446775239644       406
3  6058196446775239644       407
4  8694625920731541625       406

In [24]:
app_labels.count()

app_id      459943
label_id    459943
dtype: int64

## label_categories

In [25]:
label_categories = pd.read_csv("../datasets/label_categories.csv", sep=",")

In [26]:
label_categories.head()

label_id           category
0         1                NaN
1         2     game-game type
2         3   game-Game themes
3         4     game-Art Style
4         5  game-Leisure time

In [27]:
label_categories.count()

label_id    930
category    927
dtype: int64

In [28]:
# itt van 3 sor NaN. Igazából szerintem nem számít.

In [29]:
#stop

# Adatok előkészítése

Szerintem feladatok:
    - a gender_age_train össze kellene join-olni a phone_brand_device_modellel
    - ugyancsak a gender_age_train-t össze kell join-olni az events-app_events-app_labels -en keresztül a label_categories-vel
    - a label kategória dummy változó lenne. A problémám, hogy itt akkro  930 db lenne...
    - bemenő változók: márka, alkalmazás label dummy. Célváltozó: group


## Label counts

itt külön notebookban megszámoltam és elmentettem az egyes label-ek számosságát

In [30]:
label_categories = pd.read_csv("../datasets/label_categories_with_counts_2.csv", sep=",")

In [31]:
label_categories = label_categories.sort_values("all_cou", ascending=0)

In [32]:
label_categories["label_id_use"] = label_categories["label_id"] 

In [33]:
max_label = 30
label_array = []

In [34]:
for a in range(0,max_label):
    label_array.append("label_%d"%a)


In [35]:
for a in range(max_label, len(label_categories)):
    label_categories.at[a, 'label_id_use'] = 22222

In [36]:
# A label_id_use-ban az elso max_label helyen ott van a tenyleges id, a tobbi esetben egy gyujto: 22222

In [37]:
label_categories.head()

Unnamed: 0  label_id               category  not_active_cou  active_cou  \
0         489       548           Industry tag      14394096.0   9846501.0   
1         629       704  Property Industry 2.0      10017203.0   6807461.0   
2         490       549  Property Industry 1.0       7567242.0   6113200.0   
3         359       405           Custom label       5902239.0   2337647.0   
4         638       713             Services 1       5084162.0   3098384.0   

      all_cou  label_id_use  
0  24240597.0           548  
1  16824664.0           704  
2  13680442.0           549  
3   8239886.0           405  
4   8182546.0           713

# app_label merge

In [41]:
# az app_labels-be beleteszuk a label_id_use-t
app_labels = app_labels.merge(label_categories, on="label_id", how="left")

In [42]:
app_labels.head()

app_id  label_id  Unnamed: 0         category  not_active_cou  \
0  7324884708820027918       251         232          Finance       3518798.0   
1 -4494216993218550286       251         232          Finance       3518798.0   
2  6058196446775239644       406         360          unknown        523546.0   
3  6058196446775239644       407         361  DS_P2P net loan        523546.0   
4  8694625920731541625       406         360          unknown        523546.0   

   active_cou    all_cou  label_id_use  
0   1484275.0  5003073.0           251  
1   1484275.0  5003073.0           251  
2    204958.0   728504.0         22222  
3    204958.0   728504.0         22222  
4    204958.0   728504.0         22222

In [43]:
# a label_id_use legyen a label_id
app_labels.rename(columns={"label_id": "label_id_original", "label_id_use":"label_id"}, inplace=True)
app_labels.head()


app_id  label_id_original  Unnamed: 0         category  \
0  7324884708820027918                251         232          Finance   
1 -4494216993218550286                251         232          Finance   
2  6058196446775239644                406         360          unknown   
3  6058196446775239644                407         361  DS_P2P net loan   
4  8694625920731541625                406         360          unknown   

   not_active_cou  active_cou    all_cou  label_id  
0       3518798.0   1484275.0  5003073.0       251  
1       3518798.0   1484275.0  5003073.0       251  
2        523546.0    204958.0   728504.0     22222  
3        523546.0    204958.0   728504.0     22222  
4        523546.0    204958.0   728504.0     22222

In [45]:
events.head()

event_id            device_id            timestamp  longitude  latitude
0         1    29182687948017175  2016-05-01 00:55:25     121.38     31.24
1         2 -6401643145415154744  2016-05-01 00:54:12     103.65     30.97
2         3 -4833982096941402721  2016-05-01 00:08:05     106.60     29.70
3         4 -6815121365017318426  2016-05-01 00:06:40     104.27     23.28
4         5 -5373797595892518570  2016-05-01 00:07:18     115.88     28.66

In [47]:
app_events.head()

event_id               app_id  is_installed  is_active
0         2  5927333115845830913             1          1
1         2 -5720078949152207372             1          0
2         2 -1633887856876571208             1          0
3         2  -653184325010919369             1          1
4         2  8693964245073640147             1          1

In [48]:
stop

NameError: name 'stop' is not defined

## Tanító1 -  esemény szám + label számosság

egy olyan df-et hozunk létre, ahol minden device-hez ott van, hogy milyen típusú alkalmazás label-ek vannak, milyen a tipusu keszuleke van es hany esemenye.
Magyarázó oszlopok:
- events_num : mennyi esemenye volt
- lavel_x : az alkalmazasok (aktiv + installed kulon ) label-jei
- phone_brand
- device_model
- events_per_day : max - min timestamp osztva a napok szamaval

In [101]:
app_events.head()

event_id               app_id  is_installed  is_active
0         2  5927333115845830913             1          1
1         2 -5720078949152207372             1          0
2         2 -1633887856876571208             1          0
3         2  -653184325010919369             1          1
4         2  8693964245073640147             1          1

In [110]:
def getByDeviceID(device_id):
    df = events[ events["device_id"] == device_id ]
    df = df.merge(app_events , on="event_id", how="inner")
    df = df.merge(app_labels, on="app_id", how="left")
    return df

In [285]:
def calcRow1(device_id):
    df = getByDeviceID(device_id)
    ret = {}
    ret["device_id"] = [device_id]
    # esemenyek szama
    ret["events_num"] = df.groupby("event_id").agg("count").count().device_id 
    ret["min_timestamp"] = df['timestamp'].min()
    ret["max_timestamp"] = df['timestamp'].max()
    ret["events_per_day"] = pd.Timedelta(pd.to_datetime(ret["max_timestamp"]) - pd.to_datetime(ret["min_timestamp"])).seconds / ret["events_num"] / 60 / 60 / 24
    # label_id szamolasok
    for r in label_categories["label_id_use"].unique():
        colname = "label_%d"%r
        ret[colname] = [df[(df["label_id"] == r) & (df["is_active"] == 1)].count()["device_id"]]
        colname = "label_i_%d"%r
        ret[colname] = [df[(df["label_id"] == r) & (df["is_active"] == 0)].count()["device_id"]]
    return pd.DataFrame(data=ret)

In [ ]:
df = getByDeviceID(-6401643145415154744)
df.dtypes

In [286]:
calcRow1(6352067998666467520).head()

device_id  events_num        min_timestamp        max_timestamp  \
0  6352067998666467520          11  2016-05-03 22:14:30  2016-05-05 17:27:58   

   events_per_day  label_548  label_i_548  label_704  label_i_704  label_549  \
0         0.07282         70           34         57           19         39   

       ...        label_128  label_i_128  label_777  label_i_777  label_179  \
0      ...                0            0         11            0          4   

   label_i_179  label_787  label_i_787  label_22222  label_i_22222  
0            3         11            0          211            103  

[1 rows x 67 columns]

In [253]:
gender_age_train.head()

device_id gender  age   group
0 -8076087639492063270      M   35  M32-38
1 -2897161552818060146      M   35  M32-38
2 -8260683887967679142      M   35  M32-38
3 -4938849341048082022      M   30  M29-31
4   245133531816851882      M   30  M29-31

In [218]:
#gender_age_train_small = gender_age_train[gender_age_train["device_id"] == 6352067998666467520]
#gender_age_train_small = gender_age_train_small.append(gender_age_train[gender_age_train["device_id"] == 1390702386071991851])
#gender_age_train_small = gender_age_train.head(100) 
#gender_age_train_small.head(1000)

# csak azokkal a device_id-kal foglalkozunk, ahol vannak esemenyek

gender_age_train_filtered = gender_age_train.merge(events, on="device_id", how="inner")


In [308]:
len(gender_age_train_filtered["device_id"].unique())

23309

In [ ]:
# sorra megyunk a device_id-kon. Eleg lassu lesz
max = 100000
df_train = None
for r in gender_age_train_filtered["device_id"].unique():
    #print("Start", r)
    x = calcRow1(r)
    if df_train is None:
        df_train = x
    else:
        df_train = df_train.append(x)
    max = max -1 
    if max < 0 :
        break
df_train.head()
df_train = df_train.merge(gender_age_train, on="device_id", how="inner")
df_train = df_train.merge(phone_brand_device_model, on="device_id", how="left")
df_train.to_csv("../datasets/train1.csv")

In [290]:
df_train.head()

device_id  events_num        min_timestamp        max_timestamp  \
0 -8260683887967679142           1  2016-05-01 14:23:37  2016-05-01 14:23:37   
1  7477216237379271436           3  2016-05-04 14:08:17  2016-05-06 18:51:15   
2  6352067998666467520          11  2016-05-03 22:14:30  2016-05-05 17:27:58   
3  1508636020748379883           5  2016-05-01 12:41:38  2016-05-05 12:34:15   
4 -6876541075223249434           6  2016-05-01 02:51:41  2016-05-06 16:30:01   

   events_per_day  label_548  label_i_548  label_704  label_i_704  label_549  \
0        0.000000          5           30          3           24          3   
1        0.065502         31           29         25           17         21   
2        0.072820         70           34         57           19         39   
3        0.198975         50           39         38           36         29   
4        0.094715         63          187         46          162         48   

        ...         label_787  label_i_787  label_22222  label_i_22222  \
0       ...                 0            0            4            102   
1       ...                 3            0           27             78   
2       ...                11            0          211            103   
3       ...                 2            3           69             80   
4       ...                 0            0          126            520   

   gender  age   group  phone_brand  device_model  phone_brand_eng  
0       M   35  M32-38           小米          MI 2           Xiaomi  
1       F   37  F33-42           华为      荣耀6 plus           Huawei  
2       M   32  M32-38           华为        荣耀畅玩4X           Huawei  
3       F   28  F27-28           华为        荣耀畅玩4X           Huawei  
4       M   75    M39+           魅族        魅蓝NOTE            Meizu  

[5 rows x 73 columns]

##  SVM model
példa: https://www.datacamp.com/community/tutorials/svm-classification-scikit-learn-python

Csak numerikus magyarázó változók

In [303]:
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn import metrics


In [301]:
magyarazo_valtozok = []
for m in df_train.columns:
    if m not in ['device_id', 'min_timestamp', 'max_timestamp','gender', 'age', 'group', 'phone_brand', 'phone_brand_eng', 'device_model' ]:
        magyarazo_valtozok += [m]
#magyarazo_valtozok
X_train, X_test, y_train, y_test = train_test_split(df_train[magyarazo_valtozok], df_train['group'], test_size=0.3,random_state=109) # 70% training and 30% test


In [302]:
#Create a svm Classifier
clf = svm.SVC(kernel='linear') # Linear Kernel
#Train the model using the training sets
clf.fit(X_train, y_train)
#Predict the response for test dataset
y_pred = clf.predict(X_test)


In [304]:
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))


Accuracy: 0.22580645161290322
